lib/src/dense_set.rs

```diff
@@ -0,0 +1,1282 @@
+
+        for &item in other.reg_vec.iter() {
+            if !self.contains(item) {
+                self.reg_vec.push(item);
+            }
+        }
+
+        for i in 0..smallest_set_size {
+            self.data[i] |= other.data[i];
+        }
+
+        if other.data.len() > self.data.len() {
+            for i in smallest_set_size..greatest_set_size {
+                self.data.push(other.data[i]);
+            }
+        }
+    }
+
+    fn remove(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+
+        for &item in other.reg_vec.iter() {
+            if self.contains(item) {
+                self.vec_delete(item);
+            }
+        }
+
+        for i in 0..smallest_set_size {
+            self.data[i] &= !other.data[i];
+        }
+    }
+
+    fn intersects(&self, other: &Self) -> bool {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+
+        for i in 0..smallest_set_size {
+            if self.data[i] & other.data[i] != 0 {
+                return true;
+            }
+        }
+        false
+    }
+
+    fn is_subset_of(&self, other: &Self) -> bool {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+
+        for i in 0..smallest_set_size {
+            if (self.data[i] | other.data[i]) != other.data[i] {
+                return false;
+            }
+        }
+
+        if self.data.len() > other.data.len() {
+            for i in other.data.len()..self.data.len() {
+                if self.data[i] != 0 {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    fn equals(&self, other: &Self) -> bool {
+        self.data == other.data
+    }
+
+    fn vec_get_item(&self, index: usize, real: bool) -> Option<&Reg> {
+        for reg in self.reg_vec.iter() {
+            if reg.is_real() == real && reg.get_index() == index {
+                return Some(&reg);
+            }
+        }
+        None
+    }
+}
+
+impl fmt::Debug for BitSet {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{:?}", self.data)
+    }
+}
+
+impl Clone for BitSet {
+    fn clone(&self) -> Self {
+        BitSet {
+            data: self.data.clone(),
+            reg_vec: self.reg_vec.clone(),
+        }
+    }
+}
+
+pub struct BitSetIter<'set> {
+    set_iter: u64,
+    reg_set: &'set BitSet,
+    index: usize,
+}
+impl BitSet {
+    pub fn iter(&self) -> BitSetIter {
+        BitSetIter {
+            set_iter: self.data[0],
+            reg_set: self,
+            index: 0,
+        }
+    }
+}
+impl<'set> Iterator for BitSetIter<'set> {
+    type Item = &'set Reg;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.set_iter == 0 && self.index + 1 < self.reg_set.data.len() {
+            self.index += 1;
+            self.set_iter = self.reg_set.data[self.index];
+        }
+
+        if self.set_iter == 0 {
+            None
```

Comment:
This is incorrect: there could be a hole in the bitset, while we're not done iterating over it: think of some set which bits are like: many entries .. then enough 0 that set_iter is 0 .. then again a non-zero entry.

Can you make a test case that catches this issue, and then fix the issue, please?

Reply:
Hello @bnjbvr,
Thanks for the review!

```
        while self.set_iter == 0 && self.index + 1 < self.reg_set.bits.len() {
            self.index += 1;
            self.set_iter = self.reg_set.bits[self.index];
        }

        if self.set_iter == 0 {
            None
        }
```
It means we return `None` only  if our `vec` index is at the last element of the `vec`, and this element is equal to 0.

> Can you make a test case that catches this issue, and then fix the issue, please?

I added the following test a882811b73299cfaccf193f2cfc94e9a17041cd6 (it passes):

```
#[test]
fn hole_in_bitset() {
    let reg0 = Reg::new_real(RegClass::F64, 0, 0);
    let reg8 = Reg::new_real(RegClass::F64, 0, 8);
    let reg320 = Reg::new_virtual(RegClass::F64, 320);

    let mut a: RegBitSet = RegSet::two(reg0, reg8);
    a.insert(reg320);

    let mut a_iter = a.iter();
    assert_eq!(Some(&reg0), a_iter.next());
    assert_eq!(Some(&reg8), a_iter.next());
    assert_eq!(Some(&reg320), a_iter.next());
    assert_eq!(None, a_iter.next());
}
```

The output with (`println!("{:x?}", a.bits);`) is:
```
[101, 0, 0, 0, 0, 0, 1]
```

So  I don't think this is incorrect but I may have misunderstood your comment.